[V1T0bh] iP

V1T0bh opened the pull request.

V1T0bh committed:
Level-0 Meet Yoda
delete Duke.java
add Yoda.java

V1T0bh committed:
make code more modular
define greet and ask function

V1T0bh committed:
add ask function

V1T0bh committed:
Level-2: add list function
userInput is logged as a list that can be output

V1T0bh committed:
Level-3: add Mark and Unmark feat

add setLabel, print, mark, unmark methods in Task class
add mark and unmark actions
fix task lists

ZT712002 commented on src/main/java/Yoda.java:
```diff
@@ -0,0 +1,94 @@
+    // function to ask the user for an input
+    public static void ask() {
+        // Initialize input variable, empty task list and item count
+        final Scanner SCANNER = new Scanner(System.in);
```
Great Spot! Many people don't realize to have Scanner in full uppercase as it's a constant

ZT712002 commented on src/main/java/Yoda.java:
```diff
@@ -0,0 +1,94 @@
+         * otherwise, adds to a list */
+        String userInput =  SCANNER.nextLine();
+        while (!userInput.equals("bye")) {
+            if (userInput.equals("list")) {
+                if (count == 0) {
+                    System.out.println("This list of yours looks empty...");
+                }
+                for (int i = 0; i < count; i++) {
+                    System.out.print((i+1) + ". ");
+                    inputList[i].print();
+                }
+            } else if (userInput.split(" ")[0].equals("mark")) {
+                int itemId = Integer.parseInt(userInput.split(" ")[1]) - 1;
+                if (itemId < 0 & itemId >= count){
+                    System.out.println("Funny. This ID matches no task of yours.");
+                } else {
+                    inputList[itemId].mark();
+                }
+            } else if (userInput.split(" ")[0].equals("unmark")) {
+                int itemId = Integer.parseInt(userInput.split(" ")[1]) - 1;
+                if (itemId < 0 & itemId >= count){
+                    System.out.println("Funny. This ID matches no task of yours.");
+                } else {
+                    inputList[itemId].unmark();
+                }
+            } else {
+                System.out.print("added: ");
+                System.out.println(userInput);
+                inputList[count] = new Task(userInput);
+                count += 1;
+            }
```
Might a switch statement work here? Nothing wrong with this, just a stylistic choice

V1T0bh replied:
you're right. I'll try it out and see which works for me

ZT712002 commented on src/main/java/Yoda.java:
```diff
@@ -0,0 +1,94 @@
+                    System.out.print((i+1) + ". ");
+                    inputList[i].print();
+                }
+            } else if (userInput.split(" ")[0].equals("mark")) {
```
One suggestion here is to have a function to preprocess the user inputs, so that we can avoid redundant .split() and processing calls here.

V1T0bh replied:
that's true. I'll fix that in the next commit. thanks for the suggestion

ZT712002 reviewed (approved):
Overall good PR. With respects to coding standards there are none I've detected :) 
Most of my comments are mostly nitpicks that can be easily fixed

ZT712002 commented on src/main/java/Yoda.java:
```diff
@@ -0,0 +1,94 @@
+
+        System.out.println("What help shall I do you for?");
+
+        /* if "bye", loop terminates
```
Great comment,  having this allowed me to understand your code better

V1T0bh replied:
ikr

V1T0bh commented:
> Overall good PR. With respects to coding standards there are none I've detected :) Most of my comments are mostly nitpicks that can be easily fixed

Thank you for your suggestions! Good luck with your iP!

V1T0bh committed:
update mark and unmark function
simplified to setMark function

V1T0bh committed:
refactor code as per PR comments

change if else statemnts to switch
pre-extract first word from user input to avoid re-processing

V1T0bh committed:
implement todo, deadline, event

add processInput function to split up command into arguments
use split up arguments to assign todo, deadline, event tasks

NigelYeoTW commented on src/main/java/Yoda.java:
```diff
@@ -0,0 +1,209 @@
+        Attributes: type, label, isDone
+        methods: print, mark, unmark
+     */
+    public static class Task {
```
Perhaps create a new file for each class

NigelYeoTW commented on src/main/java/Yoda.java:
```diff
@@ -0,0 +1,209 @@
+        }
+
+        public void setLabel(String inputLabel) {
+            this.label = inputLabel;
```
Perhaps reduce the use of 'this' keyword when not needed

NigelYeoTW commented on src/main/java/Yoda.java:
```diff
@@ -0,0 +1,209 @@
+            this.label = inputLabel;
+        }
+
+        protected String taskString(){
```
Perhaps have consistent whitespace for better readability

NigelYeoTW commented on src/main/java/Yoda.java:
```diff
@@ -0,0 +1,209 @@
+        // LABEL, START, END
+        int i = 1;
+        for (int j = 1; j < 4; j ++) {
+            while (!(i >= splitArray.length || splitArray[i].contains("/"))) {
```
Perhaps you can simplify the boolean expression or extract it to make it a more readable variable

NigelYeoTW commented on src/main/java/Yoda.java:
```diff
@@ -0,0 +1,209 @@
+        String userInput =  SCANNER.nextLine();
+        while (!userInput.equals("bye")) {
+            String[] keywordsArray = processInput(userInput);
+            switch (keywordsArray[0]){
+                case "list":
```
Should there be an indentation between switch and case statements?

NigelYeoTW commented on src/main/java/Yoda.java:
```diff
@@ -0,0 +1,209 @@
+                        inputList[itemId].setMark(keywordsArray[0].equals("mark"));
+                    }
+                    break;
+                case "todo":
+                case "deadline":
+                case "event":
+                    if (count >= MAX_TASKS){
+                        System.out.println("My condolences, it seems you have too much on your plate.");
+                        System.out.println("Your task was not added.");
+                    } else {
+                        Task newTask = new Task();
+                        switch (keywordsArray[0]){
+                            case "todo":
+                                newTask = new Todo(keywordsArray[1]);
+                                break;
+                            case "deadline":
+                                newTask = new Deadline(keywordsArray[1], keywordsArray[2]);
+                                break;
+                            case "event":
+                                newTask = new Event(keywordsArray[1], keywordsArray[2], keywordsArray[3]);
+                                break;
+                        }
+                        inputList[count] = newTask;
+                        System.out.println("Successfully added: ");
+                        System.out.println(inputList[count]);
+                        count += 1;
+                    }
+                    break;
```
Perhaps you can not check for the same switch case twice to reduce nesting or consider extracting out portions of code to make it more readable

NigelYeoTW commented on src/main/java/Yoda.java:
```diff
@@ -0,0 +1,209 @@
+    public static void main(String[] args) {
+        System.out.println("------------- YODA AWAKENS -------------");
+        greet();
+        ask();
```
Perhaps the function name could be more intuitive

NigelYeoTW reviewed (commented):
Overall good attempt, there seems to be some coding standard violations, consider creating separate class files and follow the course website and coding standard website more closely to ensure further increase the quality of the code 👍

V1T0bh commented:
Thank you all for the comments. I had a sudden realiziation I PR-ed the wrong branch. SORRY! If you are seeing this, please refer to the new PR.
